src/main/scala/io/iohk/ethereum/domain/Blockchain.scala

```diff
@@ -219,8 +228,13 @@ class BlockchainImpl(
 
   override def getTotalDifficultyByHash(blockhash: ByteString): Option[BigInt] = totalDifficultyStorage.get(blockhash)
 
-  override def getBestBlockNumber(): BigInt =
-    appStateStorage.getBestBlockNumber()
+  override def getBestBlockNumber(): BigInt = {
+    val bestBlockNum = appStateStorage.getBestBlockNumber()
+    if (bestKnownBlock.get() > bestBlockNum)
```

Comment:
is there any case where this is not true?

Reply:
It is not true just after mantis start (then `bestKnownBlock` is initialized to 0) so always best-saved block number will be returned. The same case is just after fast sync finish. (fast sync do not update `bestKnownBlock` but saves all data db  and update best block number in db)

Reply:
Could this initialization in 0 cause an issue when [it's value is saved to the best block number in `appStateStorage`](https://github.com/input-output-hk/mantis/pull/411/files#diff-abb3f698c5418831dcb5f183fbf61937R383)?

Reply:
hmm it is a though question, i do not see code path that could lead to it, but theoretically it is possible.
As I think about solution to handle it more defensively, maybe it would be better to call `getBestBlockNumber()`, instead of this `bestKnownBlock.get()`. That way we would dont risk this 0 block save, and all `bestKnownBlock` hanlding would be in `getBestBlockNumber()`. Wdyt?

Reply:
Hmm that would work. Maybe we could also initialize the `bestKnownBlock` with `appStateStorage.getBestBlockNumber()`, I'd prefer that as it keeps the value `bestKnownBlock` more consistent with what it should be on initialization, wdyt?

Reply:
It sounds good but has this one caveat that we have a lot of `mock[BlockchainImpl]` in tests, and they would just throw `nullpointers` on missing `appStateStorage` during initialization. Maybe you know some work around it?

Reply:
That's a good point, there's none that I know of.
We should go with your solution then :+1: